File: app/functions/src/dto/activities.ts
import { Media, MediaJSON } from "./media";
import { Mention, MentionJSON } from "./mentions";
import { FlMeta, FlMetaJSON } from "./meta";

export const activitySchemaKey = 'activities';

/**
 * The type of activity
 * @export
 * @enum {number}
 * @property {string} Post A post
 * @property {string} Like A like
 * @property {string} Comment A comment
 * @property {string} Share A share
 * @property {string} Bookmark A bookmark
 */
export enum ActivityActionVerb {
  Post = "post",
  Like = "like",
  Comment = "comment",
  Share = "share",
  Bookmark = "bookmark",
}

/**
 * The JSON representation of an activity
 * @export
 * @interface ActivityJSON
 * @property {FlMetaJSON} [_fl_meta_]
 * @property {ActivityGeneralConfigurationJSON} [generalConfiguration]
 * @property {ActivitySecurityConfigurationJSON} [securityConfiguration]
 * @property {ActivityEventConfigurationJSON} [eventConfiguration]
 * @property {ActivityPricingInformationJSON} [pricingInformation]
 * @property {ActivityPublisherInformationJSON} [publisherInformation]
 * @property {ActivityEnrichmentConfigurationJSON} [enrichmentConfiguration]
 * @property {MediaJSON[]} [media]
 */
export interface ActivityJSON {
  _fl_meta_?: FlMetaJSON;
  generalConfiguration?: ActivityGeneralConfigurationJSON;
  securityConfiguration?: ActivitySecurityConfigurationJSON;
  eventConfiguration?: ActivityEventConfigurationJSON;
  pricingInformation?: ActivityPricingInformationJSON;
  publisherInformation?: ActivityPublisherInformationJSON;
  enrichmentConfiguration?: ActivityEnrichmentConfigurationJSON;
  media?: MediaJSON[];
}

/**
 * An activity
 * @export
 * @class Activity
 * @property {FlMeta} [_fl_meta_]
 * @property {ActivityGeneralConfiguration} [generalConfiguration]
 * @property {ActivitySecurityConfiguration} [securityConfiguration]
 * @property {ActivityEventConfiguration} [eventConfiguration]
 * @property {ActivityPricingInformation} [pricingInformation]
 * @property {ActivityPublisherInformation} [publisherInformation]
 * @property {ActivityEnrichmentConfiguration} [enrichmentConfiguration]
 * @property {Media[]} media
 */
export class Activity {
  _fl_meta_?: FlMeta;
  generalConfiguration?: ActivityGeneralConfiguration;
  securityConfiguration?: ActivitySecurityConfiguration;
  eventConfiguration?: ActivityEventConfiguration;
  pricingInformation?: ActivityPricingInformation;
  publisherInformation?: ActivityPublisherInformation;
  enrichmentConfiguration?: ActivityEnrichmentConfiguration;
  media: Media[];

  constructor(json: ActivityJSON) {
    this._fl_meta_ = json._fl_meta_ && new FlMeta(json._fl_meta_);
    this.generalConfiguration = json.generalConfiguration && new ActivityGeneralConfiguration(json.generalConfiguration);
    this.securityConfiguration = json.securityConfiguration && new ActivitySecurityConfiguration(json.securityConfiguration);
    this.eventConfiguration = json.eventConfiguration && new ActivityEventConfiguration(json.eventConfiguration);
    this.pricingInformation = json.pricingInformation && new ActivityPricingInformation(json.pricingInformation);
    this.publisherInformation = json.publisherInformation && new ActivityPublisherInformation(json.publisherInformation);
    this.enrichmentConfiguration = json.enrichmentConfiguration && new ActivityEnrichmentConfiguration(json.enrichmentConfiguration);
    this.media = json.media ? json.media.map((item) => new Media(item)) : [];
  }
}

/**
 * The type of activity
 * @export
 * @enum {number}
 * @property {string} Post A post
 * @property {string} Event An event
 * @property {string} Clip A clip
 * @property {string} Repost A repost
 */
export type ActivityGeneralConfigurationType = 'post' | 'event' | 'clip' | 'repost';

/**
 * The style of activity
 * @export
 * @enum {number}
 * @property {string} Markdown Markdown
 * @property {string} Text Text
 */
export type ActivityGeneralConfigurationStyle = 'markdown' | 'text';

/**
 * The JSON representation of an activity general configuration
 * @export
 * @interface ActivityGeneralConfigurationJSON
 * @property {ActivityGeneralConfigurationType} [type]
 * @property {ActivityGeneralConfigurationStyle} [style]
 * @property {string} [content] The content of the activity (markdown or text)
 */
export interface ActivityGeneralConfigurationJSON {
  type?: ActivityGeneralConfigurationType;
  style?: ActivityGeneralConfigurationStyle;
  content?: string;
  isSensitive?: boolean;
  repostActivityId?: string;
  repostActivityPublisherId?: string;
  repostActivityOriginFeed?: string;
}

/**
 * General configuration for an activity
 * @export
 * @class ActivityGeneralConfiguration
 * @property {ActivityGeneralConfigurationType} type
 * @property {ActivityGeneralConfigurationStyle} style
 * @property {string} content The content of the activity (markdown or text)
 */
export class ActivityGeneralConfiguration {
  type: ActivityGeneralConfigurationType;
  style: ActivityGeneralConfigurationStyle;
  content: string;
  isSensitive: boolean;
  repostActivityId: string;
  repostActivityPublisherId: string;
  repostActivityOriginFeed: string;

  constructor(json: ActivityGeneralConfigurationJSON) {
    this.type = json.type || 'post';
    this.style = json.style || 'text';
    this.content = json.content || '';
    this.isSensitive = json.isSensitive || false;
    this.repostActivityId = json.repostActivityId || '';
    this.repostActivityPublisherId = json.repostActivityPublisherId || '';
    this.repostActivityOriginFeed = json.repostActivityOriginFeed || '';
  }
}

/**
 * The mode for an activity security configuration.
 * These dictate what tag feeds the activity will appear in.
 * @export
 * @enum {number}
 * @property {string} Public Public
 * @property {string} FollowersAndConnections Followers and connections
 * @property {string} Connections Connections
 * @property {string} Private Private
 * @property {string} SignedIn signed_in
 */
export type ActivitySecurityConfigurationMode = 'public' | 'followers_and_connections' | 'connections' | 'private' | 'signed_in' | 'disabled';

/**
 * The JSON representation of an activity security configuration
 * @export
 * @interface ActivitySecurityConfigurationJSON
 * @property {string} [context]
 * @property {ActivitySecurityConfigurationMode} [viewMode]
 * @property {ActivitySecurityConfigurationMode} [reactionMode]
 * @property {ActivitySecurityConfigurationMode} [shareMode]
 */
export interface ActivitySecurityConfigurationJSON {
  context?: string;
  viewMode?: ActivitySecurityConfigurationMode;
  shareMode?: ActivitySecurityConfigurationMode;
  commentMode?: ActivitySecurityConfigurationMode;
  likesMode?: ActivitySecurityConfigurationMode;
  bookmarksMode?: ActivitySecurityConfigurationMode;
}

/**
 * Security configuration for an activity
 * @export
 * @class ActivitySecurityConfiguration
 * @property {string} context
 * @property {ActivitySecurityConfigurationMode} viewMode
 * @property {ActivitySecurityConfigurationMode} reactionMode
 * @property {ActivitySecurityConfigurationMode} shareMode
 */
export class ActivitySecurityConfiguration {
  context: string;
  viewMode?: ActivitySecurityConfigurationMode;
  shareMode?: ActivitySecurityConfigurationMode;
  commentMode?: ActivitySecurityConfigurationMode;
  likesMode?: ActivitySecurityConfigurationMode;
  bookmarksMode?: ActivitySecurityConfigurationMode;

  constructor(json: ActivitySecurityConfigurationJSON) {
    this.context = json.context || '';
    this.viewMode = json.viewMode || 'private';
    this.shareMode = json.shareMode || 'private';
    this.commentMode = json.commentMode || 'signed_in';
    this.likesMode = json.likesMode || 'public';
    this.bookmarksMode = json.bookmarksMode || 'public';
  }
}

/**
 * The JSON representation of an activity schedule
 * @export
 * @interface ActivityScheduleJSON
 * @property {string} [recurrenceRule] An IEEE RRULE string
 * @property {Date} [start] The start date of the activity
 * @property {Date} [end] The end date of the activity
 */
export interface ActivityScheduleJSON {
  recurrenceRule?: string;
  start?: Date;
  end?: Date;
}

/**
 * A schedule for an activity
 * @export
 * @class ActivitySchedule
 * @property {string} recurrenceRule An IEEE RRULE string
 * @property {Date} [start] The start date of the activity
 * @property {Date} [end] The end date of the activity
 */
export class ActivitySchedule {
  recurrenceRule: string;
  start?: Date;
  end?: Date;

  constructor(json: ActivityScheduleJSON) {
    this.recurrenceRule = json.recurrenceRule || '';
    this.start = json.start;
    this.end = json.end;
  }
}

/**
 * The JSON representation of an activity event configuration
 * @export
 * @interface ActivityEventConfigurationJSON
 * @property {Record<string, any>} [venue]
 * @property {string} [name]
 * @property {ActivityScheduleJSON} [schedule]
 * @property {string} [location]
 * @property {number} [popularityScore]
 * @property {boolean} [isCancelled]
 */
export interface ActivityEventConfigurationJSON {
  venue?: Record<string, any>;
  name?: string;
  schedule?: ActivityScheduleJSON;
  location?: string;
  popularityScore?: number;
  isCancelled?: boolean;
}

/**
 * Configuration for an activity event
 * @export
 * @class ActivityEventConfiguration
 * @property {Record<string, any>} [venue]
 * @property {string} name
 * @property {ActivitySchedule} [schedule]
 * @property {string} location
 * @property {number} popularityScore
 * @property {boolean} isCancelled
 */
export class ActivityEventConfiguration {
  venue?: any;
  name: string;
  schedule?: ActivitySchedule;
  location: string;
  popularityScore: number;
  isCancelled: boolean;

  constructor(json: ActivityEventConfigurationJSON) {
    this.venue = json.venue;
    this.name = json.name || '';
    this.schedule = json.schedule && new ActivitySchedule(json.schedule);
    this.location = json.location || '';
    this.popularityScore = json.popularityScore || 0;
    this.isCancelled = json.isCancelled || false;
  }
}

/**
 * The pricing strategy for an activity
 * @export
 * @enum {number}
 * @property {string} Persons1 Persons 1 strategy (cost per person)
 */
export type ActivityPricingExternalStoreInformationPricingStrategy = 'persons_1';

/**
 * The JSON representation of an activity pricing external store information
 * @export
 * @interface ActivityPricingExternalStoreInformationJSON
 * @property {string} [costExact]
 * @property {string} [costMinimum]
 * @property {string} [costMaximum]
 * @property {ActivityPricingExternalStoreInformationPricingStrategy} [pricingStrategy]
 */
export interface ActivityPricingExternalStoreInformationJSON {
  costExact?: string;
  costMinimum?: string;
  costMaximum?: string;
  pricingStrategy?: ActivityPricingExternalStoreInformationPricingStrategy;
}

/**
 * Pricing information for an activity
 * @export
 * @class ActivityPricingExternalStoreInformation
 * @property {string} costExact
 * @property {string} costMinimum
 * @property {string} costMaximum
 * @property {ActivityPricingExternalStoreInformationPricingStrategy} pricingStrategy
 */
export class ActivityPricingExternalStoreInformation {
  costExact: string;
  costMinimum: string;
  costMaximum: string;
  pricingStrategy: ActivityPricingExternalStoreInformationPricingStrategy;

  constructor(json: ActivityPricingExternalStoreInformationJSON) {
    this.costExact = json.costExact || '';
    this.costMinimum = json.costMinimum || '';
    this.costMaximum = json.costMaximum || '';
    this.pricingStrategy = json.pricingStrategy || 'persons_1';
  }
}

/**
 * The JSON representation of an activity pricing information
 * @export
 * @interface ActivityPricingInformationJSON
 * @property {string} [productId]
 * @property {ActivityPricingExternalStoreInformationJSON} [externalStoreInformation]
 */
export interface ActivityPricingInformationJSON {
  productId?: string;
  externalStoreInformation?: ActivityPricingExternalStoreInformationJSON;
}

/**
 * Pricing information for an activity
 * @export
 * @class ActivityPricingInformation
 * @property {string} productId
 * @property {ActivityPricingExternalStoreInformation} [externalStoreInformation]
 */
export class ActivityPricingInformation {
  productId: string;
  externalStoreInformation?: ActivityPricingExternalStoreInformation;

  constructor(json: ActivityPricingInformationJSON) {
    this.productId = json.productId || '';
    this.externalStoreInformation = json.externalStoreInformation && new ActivityPricingExternalStoreInformation(json.externalStoreInformation);
  }
}

/**
 * The JSON representation of an activity publisher information
 * @export
 * @interface ActivityPublisherInformationJSON
 */
export interface ActivityPublisherInformationJSON {
  originFeed?: string;
  publisherId?: string;
}

/**
 * Publisher information for an activity
 * @export
 * @class ActivityPublisherInformation
 */
export class ActivityPublisherInformation {
  originFeed: string;
  publisherId: string;

  constructor(json: ActivityPublisherInformationJSON) {
    this.originFeed = json.originFeed || '';
    this.publisherId = json.publisherId || '';
  }
}

/**
 * The JSON representation of an activity enrichment configuration
 * @export
 * @interface ActivityEnrichmentConfigurationJSON
 * @property {string[]} [tags]
 * @property {string} [publishLocation]
 * @property {MentionJSON[]} [mentions]
 */
export interface ActivityEnrichmentConfigurationJSON {
  tags?: string[];
  publishLocation?: string;
  mentions?: MentionJSON[];

  originFeed?: string;
  reactionCounts?: Record<string, number>;
  uniqueUserReactions?: Record<string, boolean>;
}

/**
 * Enrichment configuration for an activity
 * @export
 * @class ActivityEnrichmentConfiguration
 * @property {string[]} tags
 * @property {string} publishLocation
 * @property {Mention[]} mentions
 */
export class ActivityEnrichmentConfiguration {
  tags: string[];
  publishLocation: string;
  mentions: Mention[];
  originFeed: string;
  reactionCounts: Record<string, number>;
  uniqueUserReactions: Record<string, boolean>;

  constructor(json: ActivityEnrichmentConfigurationJSON) {
    this.tags = json.tags || [];
    this.publishLocation = json.publishLocation || '';
    this.mentions = json.mentions ? json.mentions.map((m) => new Mention(m)) : [];
    this.originFeed = json.originFeed || '';
    this.reactionCounts = json.reactionCounts || {};
    this.uniqueUserReactions = json.uniqueUserReactions || {};
  }
}

File: app/lib/dtos/database/activities/activities.dart
// Dart imports:
import 'dart:convert';

// Package imports:
import 'package:flutter_gen/gen_l10n/app_localizations.dart';
import 'package:freezed_annotation/freezed_annotation.dart';

// Project imports:
import 'package:app/dtos/converters/profile_converters.dart';
import 'package:app/dtos/database/activities/mentions.dart';
import 'package:app/dtos/database/common/fl_meta.dart';
import 'package:app/dtos/database/common/media.dart';
import 'package:app/extensions/json_extensions.dart';

part 'activities.freezed.dart';
part 'activities.g.dart';

@freezed
class Activity with _$Activity {
  const factory Activity({
    @JsonKey(name: '_fl_meta_') FlMeta? flMeta,
    @Default('') String searchDescription,
    ActivityGeneralConfiguration? generalConfiguration,
    ActivityRepostConfiguration? repostConfiguration,
    ActivitySecurityConfiguration? securityConfiguration,
    ActivityEventConfiguration? eventConfiguration,
    ActivityPricingInformation? pricingInformation,
    ActivityPublisherInformation? publisherInformation,
    ActivityEnrichmentConfiguration? enrichmentConfiguration,
    @JsonKey(fromJson: Media.fromJsonList) @Default([]) List<Media> media,
  }) = _Activity;

  factory Activity.fromJson(Map<String, dynamic> json) => _$ActivityFromJson(json);
}

@freezed
class ActivityRepostConfiguration with _$ActivityRepostConfiguration {
  const factory ActivityRepostConfiguration({
    @Default('') String targetActivityId,
    @Default('') String targetActivityPublisherId,
    @Default('') String targetActivityOriginFeed,
  }) = _ActivityRepostConfiguration;

  factory ActivityRepostConfiguration.fromJson(Map<String, dynamic> json) => _$ActivityRepostConfigurationFromJson(json);
}

@freezed
class ActivityGeneralConfiguration with _$ActivityGeneralConfiguration {
  const factory ActivityGeneralConfiguration({
    @Default(ActivityGeneralConfigurationType.post()) @JsonKey(fromJson: ActivityGeneralConfigurationType.fromJson, toJson: ActivityGeneralConfigurationType.toJson) ActivityGeneralConfigurationType type,
    @Default(ActivityGeneralConfigurationStyle.text()) @JsonKey(fromJson: ActivityGeneralConfigurationStyle.fromJson, toJson: ActivityGeneralConfigurationStyle.toJson) ActivityGeneralConfigurationStyle style,
    @Default('') String content,
    @Default(false) bool isSensitive,
  }) = _ActivityGeneralConfiguration;

  factory ActivityGeneralConfiguration.fromJson(Map<String, dynamic> json) => _$ActivityGeneralConfigurationFromJson(json);
}

@freezed
class ActivityGeneralConfigurationType with _$ActivityGeneralConfigurationType {
  const factory ActivityGeneralConfigurationType.post() = _ActivityGeneralConfigurationTypePost;
  const factory ActivityGeneralConfigurationType.event() = _ActivityGeneralConfigurationTypeEvent;
  const factory ActivityGeneralConfigurationType.clip() = _ActivityGeneralConfigurationTypeClip;
  const factory ActivityGeneralConfigurationType.repost() = _ActivityGeneralConfigurationTypeRepost;

  static String toJson(ActivityGeneralConfigurationType type) {
    return type.when(
      post: () => 'post',
      event: () => 'event',
      clip: () => 'clip',
      repost: () => 'repost',
    );
  }

  factory ActivityGeneralConfigurationType.fromJson(String value) {
    switch (value) {
      case 'post':
        return const _ActivityGeneralConfigurationTypePost();
      case 'event':
        return const _ActivityGeneralConfigurationTypeEvent();
      case 'clip':
        return const _ActivityGeneralConfigurationTypeClip();
      case 'repost':
        return const _ActivityGeneralConfigurationTypeRepost();
      default:
        throw ArgumentError('Invalid value for ActivityGeneralConfigurationType: $value');
    }
  }
}

@freezed
class ActivityGeneralConfigurationStyle with _$ActivityGeneralConfigurationStyle {
  const factory ActivityGeneralConfigurationStyle.markdown() = _ActivityGeneralConfigurationStyleMarkdown;
  const factory ActivityGeneralConfigurationStyle.text() = _ActivityGeneralConfigurationStyleText;

  static String toJson(ActivityGeneralConfigurationStyle style) {
    return style.when(
      markdown: () => 'markdown',
      text: () => 'text',
    );
  }

  factory ActivityGeneralConfigurationStyle.fromJson(String value) {
    switch (value) {
      case 'markdown':
        return const _ActivityGeneralConfigurationStyleMarkdown();
      case 'text':
        return const _ActivityGeneralConfigurationStyleText();
      default:
        throw ArgumentError('Invalid value for ActivityGeneralConfigurationStyle: $value');
    }
  }
}

@freezed
class ActivitySecurityConfiguration with _$ActivitySecurityConfiguration {
  const factory ActivitySecurityConfiguration({
    @Default('') String context,
    @Default(ActivitySecurityConfigurationMode.private()) @JsonKey(fromJson: ActivitySecurityConfigurationMode.fromJson, toJson: ActivitySecurityConfigurationMode.toJson) ActivitySecurityConfigurationMode viewMode,
    @Default(ActivitySecurityConfigurationMode.signedIn()) @JsonKey(fromJson: ActivitySecurityConfigurationMode.fromJson, toJson: ActivitySecurityConfigurationMode.toJson) ActivitySecurityConfigurationMode commentMode,
    @Default(ActivitySecurityConfigurationMode.private()) @JsonKey(fromJson: ActivitySecurityConfigurationMode.fromJson, toJson: ActivitySecurityConfigurationMode.toJson) ActivitySecurityConfigurationMode shareMode,
  }) = _ActivitySecurityConfiguration;

  factory ActivitySecurityConfiguration.fromJson(Map<String, dynamic> json) => _$ActivitySecurityConfigurationFromJson(json);
}

@freezed
class ActivitySecurityConfigurationMode with _$ActivitySecurityConfigurationMode {
  const factory ActivitySecurityConfigurationMode.public() = _ActivitySecurityConfigurationModePublic;
  const factory ActivitySecurityConfigurationMode.followersAndConnections() = _ActivitySecurityConfigurationModeFollowersAndConnections;
  const factory ActivitySecurityConfigurationMode.connections() = _ActivitySecurityConfigurationModeConnections;
  const factory ActivitySecurityConfigurationMode.private() = _ActivitySecurityConfigurationModePrivate;
  const factory ActivitySecurityConfigurationMode.signedIn() = _ActivitySecurityConfigurationModeSignedIn;
  const factory ActivitySecurityConfigurationMode.disabled() = _ActivitySecurityConfigurationModeDisabled;

  static String toJson(ActivitySecurityConfigurationMode mode) {
    return mode.when(
      public: () => 'public',
      followersAndConnections: () => 'followers_and_connections',
      connections: () => 'connections',
      private: () => 'private',
      signedIn: () => 'signed_in',
      disabled: () => 'disabled',
    );
  }

  static Iterable<ActivitySecurityConfigurationMode> values = const [
    ActivitySecurityConfigurationMode.public(),
    ActivitySecurityConfigurationMode.followersAndConnections(),
    ActivitySecurityConfigurationMode.connections(),
    ActivitySecurityConfigurationMode.private(),
    ActivitySecurityConfigurationMode.signedIn(),
    ActivitySecurityConfigurationMode.disabled(),
  ];

  static String toLocale(ActivitySecurityConfigurationMode mode, AppLocalizations localisations) {
    return mode.when(
      public: () => localisations.shared_user_type_generic_everyone,
      followersAndConnections: () => localisations.shared_user_type_generic_followers,
      connections: () => localisations.shared_user_type_generic_connections,
      private: () => localisations.shared_user_type_generic_me,
      signedIn: () => localisations.shared_user_type_generic_signed_in,
      disabled: () => localisations.shared_user_type_generic_disabled,
    );
  }

  static List<ActivitySecurityConfigurationMode> get orderedVisibilityModes {
    return <ActivitySecurityConfigurationMode>[
      const ActivitySecurityConfigurationMode.public(),
      const ActivitySecurityConfigurationMode.signedIn(),
      const ActivitySecurityConfigurationMode.connections(),
      const ActivitySecurityConfigurationMode.followersAndConnections(),
      const ActivitySecurityConfigurationMode.private(),
    ];
  }

  static String toVisibilityLocale(ActivitySecurityConfigurationMode mode, AppLocalizations localisations) {
    return mode.when(
      public: () => 'Everyone',
      followersAndConnections: () => 'Followers',
      connections: () => 'Connections',
      private: () => 'Only Me',
      signedIn: () => 'Signed in Users',
      disabled: () => 'Disabled',
    );
  }

  static List<ActivitySecurityConfigurationMode> get orderedCommentModes {
    return <ActivitySecurityConfigurationMode>[
      const ActivitySecurityConfigurationMode.public(),
      const ActivitySecurityConfigurationMode.signedIn(),
      const ActivitySecurityConfigurationMode.connections(),
      const ActivitySecurityConfigurationMode.followersAndConnections(),
      // const ActivitySecurityConfigurationMode.private(),
      const ActivitySecurityConfigurationMode.disabled(),
    ];
  }

  static String toCommentLocale(ActivitySecurityConfigurationMode mode, AppLocalizations localisations) {
    return mode.when(
      public: () => 'Everyone',
      followersAndConnections: () => 'Followers',
      connections: () => 'Connections',
      private: () => 'Only Me',
      signedIn: () => 'Signed in Users',
      disabled: () => 'Disabled',
    );
  }

  factory ActivitySecurityConfigurationMode.fromJson(String value) {
    switch (value) {
      case 'public':
        return const _ActivitySecurityConfigurationModePublic();
      case 'followers_and_connections':
        return const _ActivitySecurityConfigurationModeFollowersAndConnections();
      case 'connections':
        return const _ActivitySecurityConfigurationModeConnections();
      case 'private':
        return const _ActivitySecurityConfigurationModePrivate();
      case 'signed_in':
        return const _ActivitySecurityConfigurationModeSignedIn();
      case 'disabled':
        return const _ActivitySecurityConfigurationModeDisabled();
      default:
        throw ArgumentError('Invalid value for ActivitySecurityConfigurationMode: $value');
    }
  }
}

@freezed
class ActivityEventConfiguration with _$ActivityEventConfiguration {
  const factory ActivityEventConfiguration({
    dynamic venue,
    @Default('') String name,
    ActivitySchedule? schedule,
    @Default('') String location,
    @Default(0) int popularityScore,
    @Default(false) bool isCancelled,
  }) = _ActivityEventConfiguration;

  factory ActivityEventConfiguration.fromJson(Map<String, dynamic> json) => _$ActivityEventConfigurationFromJson(json);
}

@freezed
class ActivitySchedule with _$ActivitySchedule {
  const factory ActivitySchedule({
    @Default('') String recurrenceRule,
    DateTime? start,
    DateTime? end,
  }) = _ActivitySchedule;

  static List<ActivitySchedule> fromJsonSchedules(String data) {
    return (jsonDecode(data) as List<dynamic>).map((e) => ActivitySchedule.fromJson(json.decodeSafe(e))).toList();
  }

  factory ActivitySchedule.fromJson(Map<String, dynamic> json) => _$ActivityScheduleFromJson(json);
}

@freezed
class ActivityPricingInformation with _$ActivityPricingInformation {
  const factory ActivityPricingInformation({
    @Default('') String productId,
    ActivityPricingExternalStoreInformation? externalStoreInformation,
  }) = _ActivityPricingInformation;

  factory ActivityPricingInformation.fromJson(Map<String, dynamic> json) => _$ActivityPricingInformationFromJson(json);
}

@freezed
class ActivityPricingExternalStoreInformation with _$ActivityPricingExternalStoreInformation {
  const factory ActivityPricingExternalStoreInformation({
    @Default('') String costExact,
    @Default('') String costMinimum,
    @Default('') String costMaximum,
    @Default(ActivityPricingExternalStoreInformationPricingStrategy.onePerson()) @JsonKey(toJson: ActivityPricingExternalStoreInformationPricingStrategy.toJson, fromJson: ActivityPricingExternalStoreInformationPricingStrategy.fromJson) ActivityPricingExternalStoreInformationPricingStrategy pricingStrategy,
  }) = _ActivityPricingExternalStoreInformation;

  factory ActivityPricingExternalStoreInformation.fromJson(Map<String, dynamic> json) => _$ActivityPricingExternalStoreInformationFromJson(json);
}

@freezed
class ActivityPricingExternalStoreInformationPricingStrategy with _$ActivityPricingExternalStoreInformationPricingStrategy {
  const factory ActivityPricingExternalStoreInformationPricingStrategy.onePerson() = _ActivityPricingExternalStoreInformationPricingStrategyOnePerson;

  static String toJson(ActivityPricingExternalStoreInformationPricingStrategy strategy) {
    return strategy.when(
      onePerson: () => 'persons_1',
    );
  }

  factory ActivityPricingExternalStoreInformationPricingStrategy.fromJson(String value) {
    switch (value) {
      case 'persons_1':
        return const _ActivityPricingExternalStoreInformationPricingStrategyOnePerson();
      default:
        throw ArgumentError('Invalid value for ActivityPricingExternalStoreInformationPricingStrategy: $value');
    }
  }
}

@freezed
class ActivityPublisherInformation with _$ActivityPublisherInformation {
  const factory ActivityPublisherInformation({
    @Default('') String originFeed,
    @Default('') String publisherId,
  }) = _ActivityPublisherInformation;

  factory ActivityPublisherInformation.fromJson(Map<String, dynamic> json) => _$ActivityPublisherInformationFromJson(json);
}

@freezed
class ActivityEnrichmentConfiguration with _$ActivityEnrichmentConfiguration {
  const factory ActivityEnrichmentConfiguration({
    @JsonKey(fromJson: stringListFromJson) @Default([]) List<String> tags,
    @Default('') String promotionKey,
    @Default('') String publishLocation,
    @JsonKey(fromJson: Mention.fromJsonList, toJson: Mention.toJsonList) @Default([]) List<Mention> mentions,
  }) = _ActivityEnrichmentConfiguration;

  factory ActivityEnrichmentConfiguration.fromJson(Map<String, dynamic> json) => _$ActivityEnrichmentConfigurationFromJson(json);
}
